Implementing Shor's algorithm using the Qrisp framework on Cat Qubits.ipynb

```diff
@@ -0,0 +1,770 @@
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "963c9533",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import math\n",
+    "import random\n",
+    "import numpy as np"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c38e52c5",
+   "metadata": {},
+   "source": [
+    "We would like to perform the trivial task of prime factoring the number $15$. Everybody knows that $5\\times3 = 15$, but sometimes is better to double-check - or triple-check, as in this case.\n",
+    "\n",
+    "Why don't we start by comparing a classical brute-force period finding algorithm with its quantum counterpart and analyse their performances?\n",
+    "The implementation of this version of the Shor's algorithm is inspired by Qrisp'official tutorial."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6a4ca5aa",
+   "metadata": {},
+   "source": [
+    "#### Classical brute-force period finding\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "5c9f3942",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def is_prime(n):\n",
+    "    # returns True if n is prime\n",
+    "    if n <= 1:\n",
+    "        return False\n",
+    "    elif n <= 3:\n",
+    "        return True\n",
+    "    elif n % 2 == 0 or n % 3 == 0:\n",
+    "        return False\n",
+    "    i = 5\n",
+    "    while i * i <= n:\n",
+    "        if n % i == 0 or n % (i + 2) == 0:\n",
+    "            return False\n",
+    "        i += 6\n",
+    "    return True\n",
+    "\n",
+    "def find_period_classical(g,N):\n",
+    "    # returns the period of g in Z_N\n",
+    "    e = 1\n",
+    "    while(pow(g, e, N) != 1):\n",
+    "        e = e+1\n",
+    "    return e\n",
+    "\n",
+    "def classical_Shor(N):\n",
+    "    # returns the factors of N by using Shor's algorithm\n",
+    "    # with a classical brute-force period finding routine\n",
+    "    if N % 2 == 0:\n",
+    "        return (2, N // 2)\n",
+    "    if (is_prime(N)):\n",
+    "        return (1,N)\n",
+    "    \n",
+    "    while True:\n",
+    "        a = random.randint(2, N - 1)\n",
+    "        if math.gcd(a, N) != 1:\n",
+    "            return math.gcd(a, N), N // math.gcd(a, N)\n",
+    "        \n",
+    "        r = find_period_classical(a, N)\n",
+    "        \n",
+    "        if r % 2 == 0:\n",
+    "            x = pow(a, r // 2, N)\n",
+    "            p = math.gcd(x - 1, N)\n",
+    "            q = math.gcd(x + 1, N)\n",
+    "            if p != 1 and p != N:\n",
+    "                return (p, N // p)\n",
+    "            if q != 1 and q != N:\n",
+    "                return (q, N // q)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "e4a50a51",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Test passed!\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Test the classical Shor's algorithm for N=10000\n",
```

Comment:
I don't understand this code snippet.

Reply:
This performs the Shor's algorithm with a classical brute force period finding. Laurent's code was missing a lot of special cases for classical factorization so I implemented all of them

Reply:
Okay, seen IRL, let's remove the snippet starting this line up to line 157 👍